x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/ShapeValues.java

```diff
@@ -187,6 +200,18 @@ public double getX() throws IOException {
         public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
             throw new IllegalArgumentException("cannot write xcontent for geo_shape doc value");
         }
+
+        @Override
+        public TransportVersion getMinimalSupportedVersion() {
+            return TransportVersions.SHAPE_VALUE_SERIALIZATION_ADDED;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            byte[] bytes = this.reader.copyBytes();
```

Comment:
Does this need another copy? The `out.write()` is going to copy the bytes into the output stream again.

Reply:
Yes, the copy seems unnecessary.

Reply:
The underlying ByteArrayInputStream does not provide access to the internal `byte[]`, so if I wanted to avoid copying the bytes, I would have to edit that class to provide access, which feels risky as I assume that class does not allow this access for a reason (perhaps used in particular concurrent ways).

Reply:
Hmm I think we should make it possible for a `ByteArrayStreamInput` to copy its contents directly into a `StreamOutput`. If concurrency were a concern then it'd also be a problem in `copyBytes()`.

Reply:
I've extended the ByteArrayStreamInput with a class that implements Writeable to do this now. So I could delete copyBytes.

Reply:
Let's keep a reference to the original BytesRef, all this complexity is unecessary?